src/halmos/__main__.py

```diff
@@ -397,6 +408,279 @@ def is_global_fail_set(context: CallContext) -> bool:
     return hevm_fail or any(is_global_fail_set(x) for x in context.subcalls())
 
 
+def get_state_id(ex: Exec) -> bytes:
+    """
+    Computes the state snapshot hash, incorporating constraints on state variables.
+
+    Assumes constraints on state variables have been precomputed by running Exec.path_slice() after completing a transaction.
+    Do not use this during transaction execution.
+    """
+    return snapshot_state(ex, include_path=True).unwrap()
+
+
+def run_invariant_tests(
+    ctx: ContractContext, pre_ex: Exec, funsigs: list[str]
+) -> list[TestResult]:
+    """
+    Executes invariant test functions across multiple depths, reusing states at each depth for all invariant functions.
+
+    At each depth, starting from a given state, an arbitrary transaction is executed on every target contract, producing output states.
+    All invariant tests are then executed on each of these output states.
+    The process continues to the next depth, using these output states as the new input.
+
+    Args:
+        ctx: The context for the test contract.
+        pre_ex: The initial state from which invariant tests will be run.
+        funsigs: A list of invariant test function signatures to be executed.
+
+    Returns:
+        A list of test results.
+    """
+    args = ctx.args
+
+    # check all invariants against the initial state
+    test_results = run_tests(ctx, pre_ex, funsigs, terminal=False)
+
+    # initial test results; to be updated later
+    test_results_map = {r.name: r for r in test_results}
+
+    # Remaining tests that have not failed yet, to be executed in the next depth
+    funsigs = [r.name for r in test_results if r.exitcode == PASS]
+
+    # if no more invariant tests to run, stop
+    if not funsigs:
+        return test_results_map.values()
+
+    # dynamic set of visited states, initialized with the initial state
+    visited = set()
+    visited.add(get_state_id(pre_ex))
+
+    # dynamic list of frontier states
+    exs = [pre_ex]
+
+    # mutable context for invariant testing loop over multiple depths
+    inv_ctx = InvariantContext(
+        contract_ctx=ctx,
+        visited=visited,
+        test_results_map=test_results_map,
+    )
+
+    depth = 0
+    while True:
```

Comment:
I imagine there is something in `step_invariant_tests` that stops when the max inv-depth is reached, but wouldn't it make sense to have the condition explicitly at this outer loop?

```suggestion
    while depth <= args.inv_depth:
```

Reply:
no because different inv-depth can be given for different invariant tests.

Reply:
ah neat, let's make a note of it then

```suggestion
    # inv_depth can be overridden by specific test functions, so not checked here
    while True:
```

Reply:
is this expected?

```
    /// @custom:halmos --inv-depth 2
    function invariant_transient_storage() public {
        // note: transient storage is reset after each tx in the invariant tx sequence
        assertEq(c.tload(0), 0); // pass
    }

    /// @custom:halmos --inv-depth 4
    function invariant_transient_storage_long_running() public {
        assert(true);
    }
```

```
[PASS] invariant_transient_storage() (depth: 4, paths: 64)
[PASS] invariant_transient_storage_long_running() (depth: 4, paths: 64)
```

(the report shows them at the same depth, despite having a different config)

Reply:
it's due to a bug in printing test results. currently displaying invariant test results is messy, and needs a proper refactoring to improve, which i'd like to handle in a separate pr.

Reply:
added in the todo list